javav2/example_code/ec2/src/main/java/com/example/ec2/DescribeReservedInstances.java

```diff
@@ -23,35 +21,57 @@
-                        "state %s " +
-                        "and monitoring state %s%n",
-                    instance.reservedInstancesId(),
-                    instance.availabilityZone(),
-                    instance.instanceType(),
-                    instance.state().name());
-            });
-
-        } catch (Ec2Exception e) {
-            System.err.println(e.awsErrorDetails().errorMessage());
-            System.exit(1);
+            CompletableFuture<Void> future = describeReservedEC2InstancesAsync(ec2AsyncClient);
+            future.join(); // Wait for the async operation to complete.
+        } catch (RuntimeException rte) {
+            System.err.println("An exception occurred: " + (rte.getCause() != null ? rte.getCause().getMessage() : rte.getMessage()));
         }
     }
+
+    /**
+     * Describes the Reserved EC2 Instances asynchronously using the given {@link Ec2AsyncClient}.
+     * <p>
+     * This method uses the {@link Ec2AsyncClient#describeReservedInstances()} method to fetch the
+     * details of the Reserved EC2 Instances and prints the information about each instance to the console.
+     *
+     * @param ec2AsyncClient the {@link Ec2AsyncClient} instance to be used for the asynchronous operation
+     * @return a {@link CompletableFuture<Void>} that completes when the asynchronous operation is finished
+     */
+    public static CompletableFuture<Void> describeReservedEC2InstancesAsync(Ec2AsyncClient ec2AsyncClient) {
+        CompletableFuture<DescribeReservedInstancesResponse> response = ec2AsyncClient.describeReservedInstances();
+
+        // Handle the response or exception.
+        response.whenComplete((reservedInstancesResponse, ex) -> {
+            if (ex != null) {
+                // Handle the exception by throwing a RuntimeException
+                throw new RuntimeException("Failed to describe EC2 reserved instances.", ex);
+            } else if (reservedInstancesResponse == null || reservedInstancesResponse.reservedInstances().isEmpty()) {
+                // Throw an exception if the response is null or the result is empty
+                throw new RuntimeException("No EC2 reserved instances found.");
+            } else {
+                // Process the response if no exception occurred and the result is not empty
+                reservedInstancesResponse.reservedInstances().forEach(instance -> {
+                    System.out.printf(
+                        "Found a Reserved Instance with id %s, " +
+                            "in AZ %s, " +
+                            "type %s, " +
+                            "state %s%n",
+                        instance.reservedInstancesId(),
+                        instance.availabilityZone(),
+                        instance.instanceType(),
+                        instance.state().name());
+                });
+            }
+        });
+
+        // Return CompletableFuture<Void> to signify the async operation's completion

```

Comment:
This comment doesn't sound accurate to me. `thenApply` doesn't signify the operation's completion, it just processes results.

Reply:
checked again